app/notebook/server/NotebookManager.scala

```diff
@@ -85,82 +98,111 @@ class NotebookManager(val name: String, val notebookDir: File) {
+    val newname = getName(destPath)
+    val src = absolutePath(srcPath)
+    val dest = absolutePath(destPath)
+    val resPath = Await.result(provider.move(src, dest), DefaultOperationTimeout)
+    (newname, destPath)
   }
 
-  def save(path: String, notebook: Notebook, overwrite: Boolean) = forbidInViewer {
-    Logger.info(s"save at path $path")
+  def save(path: String, notebook: Notebook, overwrite: Boolean, message: Option[String]= None): (String, String) = forbidInViewer {
+    Logger.info(s"save at path $path with message $message")
     val file = notebookFile(path)
     if (!overwrite && file.exists()) {
-      throw new NotebookExistsException("Notebook " + path + " already exists.")
+      throw new NotebookExistsException(s"Notebook [$path] already exists.")
+    }
+    // TODO: this is expected to be sync, and the implementation is ugly...
+
+    // TODO: either `provider.save` takes an extra message parameter which changes its behavior for git (commit)
+    // or we have a dedicated function like `checkpoint` or `commit` for that purpose and `save` is just saving the file
+    val savedNb: Notebook = Await.result(provider.save(Paths.get(file.toURI), notebook, message), DefaultOperationTimeout)
+    (savedNb.metadata.get.name, path)
+  }
+
+  def checkpoints(path: String): List[Version] = {
+    Await.result(provider.versions(absolutePath(path)), DefaultOperationTimeout)
+  }
+
+  def restoreCheckpoint(path:String, id:String): Option[(String, String, String)] = {
+    val notebookCheckpoint =  provider.get(absolutePath(path), Some(Version(id,"", 1L))).map{ nb =>
+      val lastModifiedTime: String = NotebookInfo.formatTimestamp(nb.metadata.get.user_save_timestamp)
+      Some((lastModifiedTime , nb.name, path))
     }
-    FileUtils.writeStringToFile(file, Notebook.serialize(notebook), "UTF-8")
-    val nb = load(path)
-    (nb.get.metadata.get.name, path)
+    Await.result(notebookCheckpoint, DefaultOperationTimeout)
   }
 
   def load(path: String): Option[Notebook] = {
     Logger.info(s"Loading notebook at path $path")
-    val file = notebookFile(path)
-    if (file.exists())
-      Notebook.deserialize(FileUtils.readFileToString(file))
-    else
-      None
+    // TODO: this is expected to be sync, and the implementation is ugly...
+    val notebook: Future[Option[Notebook]] = provider.get(absolutePath(path)).map(Option.apply)
+    Await.result(notebook, DefaultOperationTimeout)
+      .orElse(None)
   }
 
+  protected def absolutePath(path: String): Path = {
+    Paths.get(notebookFile(path).toURI)
+  }
 
 }
 
 class NotebookExistsException(message: String) extends IOException(message)
+class DirectoryCreationException(path: File) extends Exception("Could not create dir at: " + path.getCanonicalPath)
+
+case class NotebookInfo(lastModified: String, name: String, data: String, path: String)
+object NotebookInfo {
+  val dateTimeFormat = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
```

Comment:
ohh. A shared evil `SimpleDateFormat`. This stuff is not threadsafe and will give us headaches. Consider alternatives (like jodatime or java.time) or at least instantiate it in the method below to have fresh :hankey: on each call.